test/functional/feature_index_prune.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test indices in conjunction with prune."""
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_greater_than,
+    assert_raises_rpc_error,
+)
+
+
+class FeatureIndexPruneTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 3
+        self.extra_args = [
+            ["-fastprune", "-prune=1", "-blockfilterindex=1"],
+            ["-fastprune", "-prune=1", "-coinstatsindex=1"],
+            ["-fastprune", "-prune=1", "-blockfilterindex=1", "-coinstatsindex=1"]
+        ]
+
+    def sync_index(self, height):
+        expected_filter = {
+            'basic block filter index': {'synced': True, 'best_block_height': height},
+        }
+        self.wait_until(lambda: self.nodes[0].getindexinfo() == expected_filter)
+
+        expected_stats = {
+            'coinstatsindex': {'synced': True, 'best_block_height': height}
+        }
+        self.wait_until(lambda: self.nodes[1].getindexinfo() == expected_stats)
+
+        expected = {**expected_filter, **expected_stats}
+        self.wait_until(lambda: self.nodes[2].getindexinfo() == expected)
+
+    def reconnect_nodes(self):
+        self.connect_nodes(0,1)
+        self.connect_nodes(0,2)
+        self.connect_nodes(1,2)
+        self.connect_nodes(2,1)
+
+    def mine_batches(self, n):
+        for _ in range(n):
+            self.nodes[0].generate(250)
+            self.sync_blocks()
+
+    def run_test(self):
+        filter_nodes = [self.nodes[0], self.nodes[2]]
+        stats_nodes = [self.nodes[1], self.nodes[2]]
+
+        self.log.info("check if we can access blockfilters and coinstats when pruning is enabled but no blocks are actually pruned")
+        self.sync_index(height=200)
+        tip = self.nodes[0].getbestblockhash()
+        for node in filter_nodes:
+            assert_greater_than(len(node.getblockfilter(tip)['filter']), 0)
+        for node in stats_nodes:
+            assert(node.gettxoutsetinfo(hash_type="muhash", hash_or_height=tip)['muhash'])
+
+        # Mine two batches of blocks to avoid hitting NODE_NETWORK_LIMITED_MIN_BLOCKS disconnection
+        self.mine_batches(2)
+        self.sync_index(height=700)
```

Comment:
Wouldn't it make more sense to wait for the indexes to sync *after* pruning?
Otherwise it is pretty much guaranteed that the index data at the tip (700) is available after pruning.

In general i think it would be good for the test to check for expected prune blocker heights somehow. (not sure if that is possible, i dont know that much about the functional tests)

Reply:
> Wouldn't it make more sense to wait for the indexes to sync _after_ pruning?
> Otherwise it is pretty much guaranteed that the index data at the tip (700) is available after pruning.

Yes, that would be good, but unless I am misunderstanding your comment, what we would be testing then is a race condition which would make the test flaky.

> In general i think it would be good for the test to check for expected prune blocker heights somehow. (not sure if that is possible, i dont know that much about the functional tests)

yes, I added two sanity checks for the exact prune blocker height by checking the logs